core/sql/src/main/java/org/trafodion/sql/HTableClient.java

```diff
@@ -421,29 +839,140 @@ else if (versions > 0)
+		    	  }
+		      }	    	
+		    }
+		    if (columnsToRemove != null)
+		    { //we are almost done checking if Exclude version of SingleColumnnValueFilter can be used. Th elast check s about to know if there is a IS_NULL_NULL
+		      //operation that cannot be using the Exclude method, as it is transformed in a filterList with OR, therefore we cannot guaranty that the SingleColumnValueExcludeFilter
+		      //performing the exclusion will be reached.
+		    	boolean is_null_nullFound = false;
+		    	for (Object o:compareOpList ){
+		    		if (new String((byte[])o).equals("IS_NULL_NULL")){
+		    			is_null_nullFound = true;
+		    			break;
+		    		}		    			
+		    	}
+		    	if (is_null_nullFound){
+		    		columnsToRemove = null; // disable Exclude method version of SingleColumnnValueFilter
+		    	}else
+		    		narrowDownResultColumns = false; // we will use the Exclude version of SingleColumnnValueFilter, so bypass the Family/QualifierFilter method
+		    }
+		    Filter f =constructV2Filter(colNamesToFilter,compareOpList,colValuesToCompare, columnsToRemove);
+		    if (f==null) return false; // error logging done inside constructV2Filter
+		    list.addFilter(f);
+		}//end V2
+		else{// deal with V1
+		    list = new FilterList(FilterList.Operator.MUST_PASS_ALL);
+		    
+		    for (int i = 0; i < colNamesToFilter.length; i++) {
+		      byte[] colName = (byte[])colNamesToFilter[i];
+		      byte[] coByte = (byte[])compareOpList[i];
+		      byte[] colVal = (byte[])colValuesToCompare[i];
+	
+		      if ((coByte == null) || (colVal == null)) {
+		        return false;
+		      }
+		      String coStr = new String(coByte);
+		      CompareOp co = CompareOp.valueOf(coStr);
+	
+		      SingleColumnValueFilter filter1 = 
+		          new SingleColumnValueFilter(getFamily(colName), getName(colName), 
+		              co, colVal);
+		      list.addFilter(filter1);
+		    }			
+		}//end V1
+	    // if we added a column for predicate eval, we need to filter down result columns
+	    FilterList resultColumnsOnlyFilter = null;
+	    if (narrowDownResultColumns){  	    	
+  		    HashMap<String,ArrayList<byte[]>> hm = new HashMap<String,ArrayList<byte[]>>(3);//use to deal with multiple family table
+  		    // initialize hm with list of columns requested for output
+  		    	for (int i=0; i<columns.length; i++){ // if we are here we know columns is not null
+  		    		if (hm.containsKey(new String(getFamily((byte[])columns[i])))){
+  		    			hm.get(new String(getFamily((byte[])columns[i]))).add((byte[])columns[i]);
+  		    		}else{
+  		    			ArrayList<byte[]> al = new ArrayList<byte[]>();
+  		    			al.add((byte[])columns[i]);
+  		    			hm.put(new String(getFamily((byte[])columns[i])), al);
+  		    		}	    			
+  		    	}
+  		    	
+  	    	if (hm.size()==1){//only one column family
+  	    		resultColumnsOnlyFilter = new FilterList(FilterList.Operator.MUST_PASS_ALL);
```

Comment:
Sorry, I can't find the files Qifan mentioned. What Eric mentions is my concern as well, if the two methods have similar structure and need to be synchronized, then wouldn't it be better to keep them as one method? Again, I'm saying that without seeing the new code.